[Docs] Refine the vignettes (#199)
insightsengineering/teal.transform: vignettes/data-extract-merge.Rmd

Reviewer: It seems to me the user has no influence on merging and transforming data whatsoever. The app developer specifies all operations, the can merely subset data that undergoes the transformations.

Author: Yeah, this is true. From what I see, the merge operations are just a function of using the `join_keys` to perform the joins and the app user has nothing to do with this.

Reviewer: Please don't use contractions (like "don't").
Remember to always backtick package names (here: `shiny`).
Maintain the same case across all section titles.

)
```

> `teal.transform` provides functions that help in transforming the data based on the shiny app user's input from the UI.
> The UI inputs are translated into `R` expressions to transform the data for analytical purposes.
> This gives the app user flexibility to transform the data as per their requirements.

Reviewer: ```suggestion
`teal.transform` allows the app user to oversee transforming a relational set of data objects into the final dataset for analysis.
User actions create a R expression that subsets and merges the input data.
```

> For example, let's try to create an analysis dataset `ANL` by:

Reviewer: `In the following example we will[...]`

Add somewhere: "Note that primary key columns are maintained when selecting columns from datasets."

1. Selecting the column `AGE` from `ADSL`
2. Selecting the column `AVAL` and filtering the rows where `PARAMCD` is `OS` from `ADTTE`
3. Merging the results from the above datasets using the primary keys.

<img src="images/basic_concept.svg" alt="Basic Concept of teal.transform" style="width: 100%;" />

Let's see how to achieve this dynamic `select`, `filter`, and `merge` operations in a shiny app using `teal.transform`.

> #### Step 1/5 - Creating the data extract

Reviewer: `choices = c("AGE", "BMRKR1"),` limiting choices is not connected to the essence of the example (user selecting a column). Can this be set to all columns? As a non-user: is there an easy way to say "all columns allowed"?

Author: I don't quite understand what you mean. Could you elaborate?

Reviewer: First you say 
```
1. Selecting the column `AGE` from `ADSL`
```
and then you show
```
adsl_extract <- data_extract_spec(
  dataname = "ADSL",
  select = select_spec(
    label = "Select variable:",
    choices = c("AGE", "BMRKR1"),
    selected = "AGE",
    multiple = TRUE,
    fixed = FALSE
  )
)
```

It is not obvious why I am only allowed to choose from `AGE` and `BMRKR1`.

Author: Okay, I see your point. I'll change it to `names(ADSL)` in all the places.

Author: Not in the extract vignette because we mention `select a variable from c("BMRKR1", "AGE")`

Author: But wait a minute if we do this, The UI will also have the primary keys, which will not be removed even if they stay unselected. Which might be odd. Perhaps the best thing to do is to just mention that we also limit the choices to select from. But I think it's self-explanatory. What do you think?

Reviewer: We can mention that the app dev can limit the user's choice, yes. But take care that it does not distract from the core idea of the user selecting columns to keep.

As for keys, is it possible to not select them? They are automatically included when only `AGE` is selected, aren't they?

Author: Made the `choices = c("AGE", "SEX", "BMRKR1")` for `ADSL` and `choices = c("AVAL", "AVALC", "ASEQ")` for `ADTTE` as the diagram simulates those are the available columns before the selection.

Author: Irrespective of the user selecting the key columns they will be selected by `teal.transform` so this feels like a bug for the app user.
I will add a comment about the ability to restrict the option choices, as it's an important feature to showcase.

Author: How does it look now?
 
<img width="757" alt="Screenshot 2024-02-16 at 8 00 38 PM" src="https://github.com/insightsengineering/teal.transform/assets/49812166/a0b745ea-9e95-4094-900e-a01101d58d4f">

> #### Step 1/5 - Creating the data extract

Reviewer: ```suggestion
#### Step 1/5 - Creating data extract specifications
```

What is created is extract _specs_, not data extracts (those will be created in the srv functions and merged). Consider changing variable names accordingly.

)
```

_Note that `teal.transform` is about to undergo a refactor so the functions in this vignette are subject to change._

`teal.transform` provides functions that help in transforming the data based on the shiny app user's input from the UI.
The UI inputs are translated into `R` expressions to transform the data for analytical purposes.
This gives the app user flexibility to transform the data as per their requirements.

For example, let's try to create an analysis dataset `ANL` by:

1. Selecting the column `AGE` from `ADSL`
2. Selecting the column `AVAL` and filtering the rows where `PARAMCD` is `OS` from `ADTTE`
3. Merging the results from the above datasets using the primary keys.

<img src="images/basic_concept.svg" alt="Basic Concept of teal.transform" style="width: 100%;" />

Let's see how to achieve this dynamic `select`, `filter`, and `merge` operations in a shiny app using `teal.transform`.

> #### Step 1/5 - Creating the data extract

Reviewer 2: I like the flow of this document :+1:

data_extracts <- list(adsl_extract = adsl_extract, adtte_extract = adtte_extract)
```

#### Step 2/5 - Creating the UI

Here, we define the `merge_ui` function, which will be used to create the UI components for the `shiny` app.

Note that we take in the list of `data_extract` objects as input, and make use of the `data_extract_ui` function to create our UI.

```{r}
merge_ui <- function(id, data_extracts) {
  ns <- NS(id)
  sidebarLayout(
    sidebarPanel(
      h3("Encoding"),
      div(
        data_extract_ui(
          ns("adsl_extract"), # must correspond with data_extracts list names
          label = "ADSL extract",
          data_extracts[[1]]
        ),
        data_extract_ui(
          ns("adtte_extract"), # must correspond with data_extracts list names
          label = "ADTTE extract",
          data_extracts[[2]]
        )
      )
    ),
    mainPanel(
      h3("Output"),
      verbatimTextOutput(ns("expr")),
      dataTableOutput(ns("data"))
    )
  )
}
```

#### Step 3/5 - Creating the Server Logic

Here, we define the `merge_srv` function, which will be used to create the server logic for the `shiny` app.

> This function gets the `datasets` as a list of reactive `data.frame`, the list of `data_extract` objects, and the `join_keys` object as input.

Reviewer: ```suggestion
This function takes as arguments the datasets (as a list of reactive `data.frame`), the data extract specifications created above (the `data_extract` list), and the `join_keys` object.
```

`join_keys` has not been mentioned before. Add some explanation somewhere or at least change `the` to `a` and add a link to `teal.data` here.

```{r}
merge_ui <- function(id, data_extracts) {
  ns <- NS(id)
  sidebarLayout(
    sidebarPanel(
      h3("Encoding"),
      div(
        data_extract_ui(
          ns("adsl_extract"), # must correspond with data_extracts list names
          label = "ADSL extract",
          data_extracts[[1]]
        ),
        data_extract_ui(
          ns("adtte_extract"), # must correspond with data_extracts list names
          label = "ADTTE extract",
          data_extracts[[2]]
        )
      )
    ),
    mainPanel(
      h3("Output"),
      verbatimTextOutput(ns("expr")),
      dataTableOutput(ns("data"))
    )
  )
}
```

#### Step 3/5 - Creating the Server Logic

Here, we define the `merge_srv` function, which will be used to create the server logic for the `shiny` app.

This function gets the `datasets` as a list of reactive `data.frame`, the list of `data_extract` objects, and the `join_keys` object as input.
We make use of the `merge_expression_srv` function to get a reactive list containing merge expression and information needed to perform the transformation - see more in `merge_expression_srv` documentation.

> We render this expression and evaluate it to get the final `ANL` dataset which is also rendered as a table in our UI.

Reviewer: What do you mean by rendering an expression?

Author: ```suggestion
We evaluate this expression to get the final `ANL` dataset which is also rendered as a table in our UI.
```

How about now?

Author: No wait. I remember what I meant. Let me rephrase it.

Author: ```suggestion
We print this expression in the UI  and also evaluate it to get the final `ANL` dataset which is also displayed as a table in the UI.
```

Is this clear now? I wanted to convey that this object is now used to create two outputs in the UI, one is the print output of the expression and the other is the table. It would be easier to see this once we add a screenshot.

Author: This is just explaining what we're doing in the example app created in the vignette. Not the logic behind what is going on. I've added the screenshot of the app now. Does it make things clear now?

Reviewer: Oooh, I see. That's better.

}
```

#### Step 4/5 - Preparing the Data

> The `data_extract_srv` module depends on a list of reactive or non-reactive `data.frame` objects. 

Reviewer: ```suggestion
The `data_extract_srv` module takes as argument a list of reactive or non-reactive `data.frame`s. 
```

> The `data_extract_srv` module depends on a list of reactive or non-reactive `data.frame` objects. 

Reviewer: `data_extract_srv` is not called in this section so I'm not sure the flow is correct.

Maybe preparing data and joining keys should be step 1? It would be a good place to mention `join_keys`.

Author: It should be `merge_expression_srv`.
The `data_extract_srv`, `merge_expression_srv`, and `merge_expression_module` can be used in the module server along with `data_extract_ui` in the module UI.

Author: I'll remove it here. Looks like it was a copy paste from the other vignette. We already explained about `merge_expression_srv` above.

Author: Removed in both the places. They are completely redundant.

)
```

#### Step 5/5 - Creating the Shiny App

> Finally, we include `merge_ui` and `merge_srv` in the UI and server components of the `shinyApp`, respectively,
> using the `data_extract`s defined in the first code block and the `datasets` object:

Reviewer: ```suggestion
Finally, we include appropriate `merge_ui` and `merge_srv` calls in the UI and server of our `shiny` app, respectively.
```
